src/main/scala/devices/debug/Debug.scala

```diff
@@ -289,10 +297,103 @@ class TLDebugModuleOuter(device: Device)(implicit p: Parameters) extends LazyMod
+    val HAWINDOWSELRdEn   = Wire(init = false.B)
+
+    val HAWINDOWRdData = Wire(init = (new HAWINDOWFields()).fromBits(0.U))
+    val HAWINDOWWrDataVal = Wire(init = 0.U(32.W))
+    val HAWINDOWWrData = (new HAWINDOWFields()).fromBits(HAWINDOWWrDataVal)
+    val HAWINDOWWrEn   = Wire(init = false.B)
+    val HAWINDOWRdEn   = Wire(init = false.B)
+
+    if (supportHartArray) {
+      val HAWINDOWSELReset = Wire(init = (new HAWINDOWSELFields().fromBits(0.U)))
+      val HAWINDOWSELNxt = Wire(init = (new HAWINDOWSELFields().fromBits(0.U)))
+      val HAWINDOWSELReg = Wire(init = new HAWINDOWSELFields().fromBits(AsyncResetReg(updateData = HAWINDOWSELNxt.asUInt,
+        resetData = 0,
+        enable = true.B,
+        name = "HAWINDOWSELReg"
+      )))
+
+      HAWINDOWSELRdData := Wire(init = HAWINDOWSELReg)
+      HAWINDOWSELNxt := HAWINDOWSELReg
+      when (~dmactive) {
+        HAWINDOWSELNxt := HAWINDOWSELReset
+      } .otherwise {
+        when (HAWINDOWSELWrEn) {
+            // Unneeded upper bits of HAWINDOWSEL are tied to 0.  Entire register is 0 if all harts fit in one window
+          if (nComponents > haWindowSize) {
+            HAWINDOWSELNxt.hawindowsel := HAWINDOWSELWrData.hawindowsel & (log2Up(nComponents) - 5).U
+          } else {
+            HAWINDOWSELNxt.hawindowsel := 0.U
+          }
+        }
+      }
+
+      val numHAMASKSlices = ((nComponents - 1)/haWindowSize)+1
+      HAWINDOWRdData.maskdata := 0.U     // default, overridden below
+      for (ii <- 0 until numHAMASKSlices) {
+        val sliceMask = if (nComponents > ((ii*haWindowSize) + haWindowSize-1)) 0xFFFFFFFF  // All harts in this slice exist
+                        else (1<<(nComponents - (ii*haWindowSize))) - 1         // Partial last slice
+        val HAMASKRst = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKNxt = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKReg = Wire(init = Vec(AsyncResetReg(updateData = HAMASKNxt.asUInt,
+          resetData = 0,
+          enable = true.B,
+          name = s"HAMASKReg${ii}")))
+
+        when (ii.U === HAWINDOWSELReg.hawindowsel) {
+          HAWINDOWRdData.maskdata := HAMASKReg.asUInt & sliceMask.U
+        }
+
+        HAMASKNxt.maskdata := HAMASKReg.asUInt
+        when (~dmactive) {
+          HAMASKNxt := HAMASKRst
+        }.otherwise {
+          when (HAWINDOWWrEn && (ii.U === HAWINDOWSELReg.hawindowsel)) {
+            HAMASKNxt.maskdata := HAWINDOWWrData.maskdata
+          }
+        }
+
+         // drive each slice of hamask with stored HAMASKReg or with new value being written
+        for (jj <- 0 until haWindowSize) {
+          if (((ii*haWindowSize) + jj) < nComponents) {
+            val tempWrData = HAWINDOWWrData.maskdata.toBools
+            val tempMaskReg = HAMASKReg.asUInt.toBools
+            hamask((ii*haWindowSize) + jj) := Mux(HAWINDOWWrEn && (ii.U === HAWINDOWSELReg.hawindowsel), tempWrData(jj), tempMaskReg(jj))
```

Comment:
this is fine/correct, but you could perhaps more clearly write it with a `when`:

```
when (HAWINDOWWrEn && (ii.U ...) {
  hamask((ii*haWindowSize) + jj) := tempWrData(jj)
}
```